In [1]:
from pathlib import Path
import pandas as pd

DATA_PATH = Path("../data/processed/loan_default_final_outcomes.csv")

df = pd.read_csv(DATA_PATH)
df.shape

(1348099, 18)

In [3]:
print("Rows and coumns:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

print("\nData types and non-null counts:")
df.info()

Rows and coumns: (1348099, 18)

Column names:
['id', 'loan_amnt', 'term_months', 'int_rate', 'grade', 'sub_grade', 'emp_length_years', 'home_ownership', 'annual_inc', 'verification_status', 'issue_date', 'issue_year', 'issue_month', 'loan_status', 'outcome_group', 'default_flag', 'purpose', 'dti']

Data types and non-null counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1348099 entries, 0 to 1348098
Data columns (total 18 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   id                   1348099 non-null  int64  
 1   loan_amnt            1348099 non-null  float64
 2   term_months          1348099 non-null  int64  
 3   int_rate             1348099 non-null  float64
 4   grade                1348099 non-null  object 
 5   sub_grade            1348099 non-null  object 
 6   emp_length_years     1269549 non-null  float64
 7   home_ownership       1348099 non-null  object 
 8   annual_inc           1348095 n

In [5]:
missing_summary = (
    df.isna()
    .sum()
    .to_frame("missing_count")
    .assign(
        missing_percentage = lambda table:
        (table["missing_count"] / len(df) * 100).round(2)
    )
    .sort_values("missing_count", ascending=False)
)

display(missing_summary)

print("Duplicate rows:", df.duplicated().sum())
print("Duplicated loan IDs:", df["id"].duplicated().sum())

,missing_count,missing_percentage
emp_length_years,78550,5.83
dti,376,0.03
annual_inc,4,0.00
issue_date,0,0.00
purpose,0,0.00
default_flag,0,0.00
outcome_group,0,0.00
loan_status,0,0.00
issue_month,0,0.00
issue_year,0,0.00


Duplicate rows: 0
Duplicated loan IDs: 0


In [12]:
print("Outcome groups:")
display(df["outcome_group"].value_counts(dropna=False))

print("Default_flag:")
display(df["default_flag"].value_counts(dropna=False))

print("Default_rate:")
print(df["default_flag"].mean().round(4) * 100)

Outcome groups:


outcome_group
final_good    1078739
final_bad      269360
Name: count, dtype: int64

Default_flag:


default_flag
0.0    1078739
1.0     269360
Name: count, dtype: int64

Default_rate:
19.98


In [14]:
display(
    pd.crosstab(
        df["outcome_group"],
        df["default_flag"],
        margins=True
    )
)

default_flag,0.0,1.0,All
outcome_group,,,
final_bad,0,269360,269360
final_good,1078739,0,1078739
All,1078739,269360,1348099


In [20]:
display(df.head(3))

display(
    df[
        [
            "loan_amnt",
            "term_months",
            "int_rate",
            "grade",
            "annual_inc",
            "dti",
            "outcome_group",
            "default_flag",
        ]
    ].describe(include="all").T
)

,id,loan_amnt,term_months,int_rate,grade,sub_grade,emp_length_years,home_ownership,annual_inc,verification_status,issue_date,issue_year,issue_month,loan_status,outcome_group,default_flag,purpose,dti
0,68407277,3600.0,36,13.99,C,C4,10.0,MORTGAGE,55000.0,Not Verified,2015-12-01,2015,12,Fully Paid,final_good,0.0,debt_consolidation,5.91
1,68355089,24700.0,36,11.99,C,C1,10.0,MORTGAGE,65000.0,Not Verified,2015-12-01,2015,12,Fully Paid,final_good,0.0,small_business,16.06
2,68341763,20000.0,60,10.78,B,B4,10.0,MORTGAGE,63000.0,Not Verified,2015-12-01,2015,12,Fully Paid,final_good,0.0,home_improvement,10.78


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
loan_amnt,1348099.0,NaN,NaN,NaN,14408.998913,8716.137925,500.0,7975.0,12000.0,20000.0,40000.0
term_months,1348099.0,NaN,NaN,NaN,41.78402,10.264584,36.0,36.0,36.0,36.0,60.0
int_rate,1348099.0,NaN,NaN,NaN,13.241562,4.765685,5.31,9.75,12.74,15.99,30.99
grade,1348099,7,B,393102,NaN,NaN,NaN,NaN,NaN,NaN,NaN
annual_inc,1348095.0,NaN,NaN,NaN,76237.743295,69922.741975,0.0,45750.0,65000.0,90000.0,10999200.0
dti,1347723.0,NaN,NaN,NaN,18.274282,11.155478,0.0,11.79,17.61,24.05,999.0
outcome_group,1348099,2,final_good,1078739,NaN,NaN,NaN,NaN,NaN,NaN,NaN
default_flag,1348099.0,NaN,NaN,NaN,0.199807,0.399856,0.0,0.0,0.0,0.0,1.0


## Day 1 Audit Summary

- The dataset contains 1348099 loan records and 18 columns.
- Each row represents one loan, so the unit of analysis is loan-level.
- The dataset contains 0 duplicate rows and 0 duplicate loan IDs.
- The column with the highest missingness is `emp_length_years`, with 5.83%.
- The target contains 1078739 good loans and 269360 bad loans.
- The overall bad-loan rate is 19.98%.
- The dataset is imbalanced because the good loans are much more common than bad loans.
- The target fields `outcome_group` and `default_flag` agree fully.
- `dti` has 376 missing values and `annual_inc` has 4 missing values. 